rfcs/026-work-enricher/README.md

```diff
@@ -0,0 +1,160 @@
+
+* The fact that the relations are not indexed up front means we can not use related data within search queries (such as if we wanted to include the title of the root of the collection for any descendent works, for example).
+
+* The use of paths for encoding the structure of the tree couples relations tightly to Calm. There is upcoming use cases for relations beyond just archives, and on a conceptual level the goal of the pipeline should be to generate data which is abstracted away from any particular source system.
+
+We therefore need a method to generate relation data on works within the pipeline, in advance of indexing for the API.
+
+## Enricher
+
+In order to include relation data on the work model itself it is clear there needs to be an additional stage of the pipeline which is able to query relationships between works, and include data from related works onto the work itself. This proposed name for this stage is the enricher.
+
+It is important that this stage takes place after the merger: if denormalisation happens before this stage the related data may be inconsistent with what exists on the work post merging. For this reason the enricher should exist near the end of the pipeline immediately preceding the ID minter.
+
+There are a few possible methods by which the enricher can query relationships between works:
+
+1) Assuming the merged works coming into the enricher are stored within some Elasticsearch index, the existing API code for querying on the archive path could be repurposed for this.
+
+2) Before coming into the enricher we could build and store a directed graph representation of the archives in some data store, with each node being a work and the edges encoding relations. We will discuss this further in the [Graph Store](#graph-store) section.
+
+In either case, denormalisation of works within the enricher would consist of:
+
+1. Fetching the relations for a given work
+2. Emitting a new work containing the relevant fields embedded (`parts`, `partOf`, `precededBy`, `succeededBy`)
+3. Traversing the relations and emitting works for these in turn.
+
+Part `3` here is required as we have two way relations encoded for particular fields: for example, if in the tree below there is a node added with path `A/1/Z`, we not only need to send this work downstream but also send work `A/1` (containing updated `parts`), and works `A/1/X` and `A/1/Y` (containing updated `succeededBy`).
+
+```
+A
+├── 1
+│   ├── X
+│   └── Y
+│
+└── 2
+```
+
+Generally, for the insertion, update or removal of any given node, we need to update its siblings, its parent, and the whole subtree containing its descendants. The worst case for this process is an update of the root node, in which case the whole archive needs to be denormalised, with the largest archives containing in the order of 1000 works.
```

Comment:
I can see why insertion or removal triggers a cascade of related works – why is that true for updates? Doesn't it depend on what part of the work was updated? e.g. if a field that's not on the partial works get updated, none of the related works would actually change

Reply:
Yes that is true, I mention that in the problems section at the bottom